### ISSUE TITLE
The :f adverb in quoted strings may misbehave.

### ISSUE THREAD

**Issue by lucs** [rakudo/rakudo#5646]
Given:

    my $s = 'baz'
    sub foo { $^a }

compare these expressions:

    Q:c'{$s}<jub>'
    Q:f'&foo($s)<jub>'

I believe both should evaluate to `baz<jub>`, but the second one fails with `Type Str does not support associative indexing`.

**Comment by lizmat** [comment]
Disagree that this is an error, but a misunderstanding of what interpolation does (and possibly a deficiency in the documentation @coke).

Let's take this example:
```raku
my $a = *.say;
say Q:s/$a("foo")/;
```
Do you expect this to output the string representation of `*.say` (so: `WhateverCode.new("foo")␤`)?   Or do you expect the WhateverCode to be executed, and get `foo␤True␤` ?

The latter is what happens, and what I think people would expect.

The same applies to your example: `Q:f'&foo($s)<jub>`.

**Comment by lucs** [comment]
I agree that I would expect your example to work as it does.

So is the idea that the :s, :a, :h, and :f adverbs consider postcircumfix operators to apply, but not the :c one?

**Comment by lizmat** [comment]
The `:c` one creates a scope from which anything can be returned.  And whatever is being returned, is stringified: `Q:c'{$s}<jub>'` basically codegens to `{ $s }() ~ <jub>` (note that the stringification is implicit by the `infix:<~>` op).  Whereas `Q:s'foo$s<jub>` codegens to `"foo" ~ $s<jub>`.

Note that all of this happens at compilation by altering the quoting slang from the given adverbs for the given string.

Hope this made sense  :-)

**Comment by lucs** [comment]
Yep, made sense, and made me a bit sad. I use Raku to generate LaTeX files so I have ``sub f { $^a }``and can do stuff like ``my $latex = Q:f'\baz{ bla &f(<b l a>.join) bla }'``. But that fails if it is ``\baz{ &f($foo.uc)<jub> }`` for example. Using instead the ``:c`` adverb can solve that, but I'd then have tons of opening brackets to escape in the main LaTeX text.

I can work around that by using custom evaluate-this-as-a-closure-delimiters, something like `` \baz{ ‹$foo.uc›<jub> }`` and search/replacing them myself.

What would be really nice is some mechanism to invoke the ``:c`` adverb with a different set of enclosing delimiters. Is such a thing feasible?

**Comment by lizmat** [comment]
> What would be really nice is some mechanism to invoke the :c adverb with a different set of enclosing delimiters. Is such a thing feasible?

It's a SMOP.

**Comment by lizmat** [comment]
However, there may be a better way around this.  Make the `f` subroutine return a special version of `Str`:
```raku
class S is Str {
    method AT-KEY($a) {
        dd $a;
        self ~ "<" ~ $a ~ ">"
    }
}
sub f($a) { S.new(value => $a) }
say "before &f("foo")<bar> after";  # before foo<bar> after
```

**Comment by lucs** [comment]
That ``class S`` solution is very interesting, but it appears that the ``AT-KEY`` method's argument is what was enclosed by the pointy brackets, but trimmed of whitespace. So ``"&f("foo")< bar >"`` obtains ``"foo<bar>"`` rather than the desired ``foo< bar >``, and it would fail to compile given for example ``"&f("foo")<- "``.
